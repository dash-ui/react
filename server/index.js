'use strict'

module.exports = require('@-ui/styles/server')
