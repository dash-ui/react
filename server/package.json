{
  "name": "@-ui/react-server",
  "version": "1.0.0",
  "homepage": "https://github.com/dash-ui/react/tree/master/server#readme",
  "repository": "github:dash-ui/react",
  "bugs": "https://github.com/dash-ui/react/issues",
  "license": "MIT",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "source": "src/index.tsx",
  "types": "src/index.d.ts",
  "files": [
    "/src",
    "/dist"
  ],
  "sideEffects": false
}
