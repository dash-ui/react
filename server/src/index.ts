import * as React from 'react'
import {renderToStaticMarkup} from 'react-dom/server'
import defaultStyles from '@dash-ui/styles'
import {
  createStylesFromString,
  CreateStylesOptions,
} from '@dash-ui/styles/server'
import type {Styles} from '@dash-ui/styles'

export const toComponent = (
  html: string,
  styles: Styles = defaultStyles,
  options?: CreateStylesOptions
): React.ReactElement => {
  const {dash} = styles
  const {names, css} = createStylesFromString(html, styles, options)
  return React.createElement('style', {
    key: dash.key,
    nonce: dash.sheet.nonce ? dash.sheet.nonce : void 0,
    'data-dash': names.join(' '),
    'data-cache': dash.key,
    dangerouslySetInnerHTML: {__html: css},
  })
}

export interface StyleProps {
  html: string
  styles?: Styles
  options?: CreateStylesOptions
}

export const Style: React.FC<StyleProps> = ({html, styles, options}) =>
  toComponent(html, styles, options)

/* istanbul ignore next */
export const createGatsbyRenderer = (
  styles: Styles,
  options?: CreateStylesOptions
) => {
  return function replaceRenderer(props: any): any {
    const bodyString = renderToStaticMarkup(props.bodyComponent)
    props.setHeadComponents([toComponent(bodyString, styles, options)])
    return props
  }
}

export * from '@dash-ui/styles/server'
