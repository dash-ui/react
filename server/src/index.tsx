/* eslint-disable @typescript-eslint/no-var-requires */
import { styles as defaultStyles } from "@dash-ui/styles";
import type { Styles } from "@dash-ui/styles";
import { createStylesFromString } from "@dash-ui/styles/server";
import * as React from "react";

/**
 * A function for creating a React `<style>` component for
 * inserting Dash styles in SSR.
 *
 * @param html - The HTML generated by `renderToStaticMarkup()` or `renderToString()`
 * @param styles - An instance of `styles()`. Defaults to the default styles instance
 *   in `@dash-ui/styles`.
 */
export function toComponent(
  html: string,
  styles: Styles<any, any> = defaultStyles
): React.ReactElement {
  const { dash } = styles;
  const { names, css } = createStylesFromString(html, styles);

  return (
    <style
      key={dash.key}
      nonce={dash.sheet.nonce ? dash.sheet.nonce : void 0}
      data-dash={names.join(" ")}
      data-cache={dash.key}
      dangerouslySetInnerHTML={{ __html: css }}
    />
  );
}

/**
 * A React component for injecting SSR CSS styles into Next.js documents
 *
 * @param root0
 * @param root0.html
 * @param root0.styles
 * @example
 * // _document.js
 * import React from 'react'
 * import Document from 'next/document'
 * import {Style} from '＠dash-ui/react/server'
 *
 * export default class MyDocument extends Document {
 *   static async getInitialProps(ctx) {
 *     const initialProps = await Document.getInitialProps(ctx)
 *     return {
 *      ...initialProps,
 *      styles: (
 *         <>
 *           {initialProps.styles}
 *           <Style html={initialProps.html}/>
 *         </>
 *       ),
 *     }
 *   }
 * }
 */
export function Style({ html, styles }: StyleProps) {
  return toComponent(html, styles);
}

export interface StyleProps {
  /**
   * The HTML generated by Next.js, `renderToStaticMarkup()` or `renderToString()`
   */
  html: string;
  /**
   * An instance of `styles()`. Defaults to the default styles instance in `@dash-ui/styles`.
   */
  styles?: Styles<any, any>;
}

/**
 * Creates a Gatsby replaceRenderer for injecting styles generated by Dash on
 * the server into the Gatsby `<head>` component
 *
 * @param styles - An instance of `styles()`. Defaults to the default styles instance
 *   in `@dash-ui/styles`.
 * @example
 * // gatsby-ssr.js
 * exports.replaceRenderer = require('＠dash-ui/react/server').createGatsbyRenderer()
 */
export function createGatsbyRenderer(styles: Styles<any, any> = defaultStyles) {
  /* istanbul ignore next */
  return function replaceRenderer<P = any>(props: P): P {
    const bodyString = require("react-dom/server").renderToStaticMarkup(
      // @ts-expect-error
      props.bodyComponent
    );
    // @ts-expect-error
    props.setHeadComponents([toComponent(bodyString, styles)]);
    return props;
  };
}

export * from "@dash-ui/styles/server";
