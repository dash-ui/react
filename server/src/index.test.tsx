/**
 * @jest-environment node
 */
import * as React from 'react'
import {renderToStaticMarkup} from 'react-dom/server'
import styles, {createStyles} from '@dash-ui/styles'
import {Style} from './index'

describe('<Style>', () => {
  it('creates <style> tag based on html string', () => {
    const styles = createStyles()
    const style = styles({
      default: 'display: block;',
    })
    const Component = () => <div className={style()} />

    const result = renderToStaticMarkup(
      <Style html={renderToStaticMarkup(<Component />)} styles={styles} />
    )

    expect(result).toMatchSnapshot()
  })

  it('creates <style> tag with nonce', () => {
    const styles = createStyles({nonce: 'abc'})
    const style = styles({
      default: 'display: block;',
    })
    const Component = () => <div className={style()} />
    const result = renderToStaticMarkup(
      <Style html={renderToStaticMarkup(<Component />)} styles={styles} />
    )

    expect(result).toMatchSnapshot()
  })

  it('creates <style> tag with cache key', () => {
    const styles = createStyles({key: 'abc'})
    const style = styles({
      default: 'display: block;',
    })
    const Component = () => <div className={style()} />
    const result = renderToStaticMarkup(
      <Style html={renderToStaticMarkup(<Component />)} styles={styles} />
    )

    expect(result).toMatchSnapshot()
  })

  it('creates <style> tag with default styles function', () => {
    const style = styles({
      default: 'display: block;',
    })
    const Component = () => <div className={style()} />
    const result = renderToStaticMarkup(
      <Style html={renderToStaticMarkup(<Component />)} />
    )

    expect(result).toMatchSnapshot()
  })
})
