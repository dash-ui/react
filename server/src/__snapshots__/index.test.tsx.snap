// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Style> creates <style> tag based on html string 1`] = `"<style data-dash=\\"1abu38z\\" data-cache=\\"ui\\">.ui-1abu38z{display:block;}</style>"`;

exports[`<Style> creates <style> tag with cache key 1`] = `"<style data-dash=\\"1abu38z\\" data-cache=\\"abc\\">.abc-1abu38z{display:block;}</style>"`;

exports[`<Style> creates <style> tag with default styles function 1`] = `"<style data-dash=\\"1abu38z\\" data-cache=\\"ui\\">.ui-1abu38z{display:block;}</style>"`;

exports[`<Style> creates <style> tag with nonce 1`] = `"<style nonce=\\"abc\\" data-dash=\\"1abu38z\\" data-cache=\\"ui\\">.ui-1abu38z{display:block;}</style>"`;
